src/libcore/iter/range.rs

```diff
@@ -311,7 +549,13 @@ impl<A: Step> Iterator for ops::RangeFrom<A> {
 
     #[inline]
     fn next(&mut self) -> Option<A> {
```

Comment:
`RangeFrom` is a very tricky case (cc #25708, @ollie27, https://github.com/rust-lang/rust/pull/62886#discussion_r314946970).

Because of the [stable declaration of the type](https://doc.rust-lang.org/stable/std/ops/struct.RangeFrom.html), it is physically impossible to store the e.g. 257 states for `RangeFrom` to yield `0, 1, 2, ..., 254, 255, panic, ...`. The current implementation (and the one here) in debug mode yield `253, 254, panic, ...` and in release mode yield `254, 255, 0, 1, ...`. Mixing in `nth` with `next` calls makes it even more interesting: `nth` always panics if it would overflow `u8`, and when it should return 255, in release mode sets the range to `0..`, and in debug mode panics. So `nth` is not consistent with `n` calls to `next` due to this behavior of only sometimes panicking when overflowing.

This PR shouldn't change the observable behavior of `RangeFrom`, but it's worth discussing where it is desired to end up with `RangeFrom`, as that will shape what behaviors `Step` needs to provide. But no matter what we choose, `n..` must behave differently from `n..MAX`.

The four consistent potential solutions are:

- Unconditional saturation: `0, 1, ..., 254, 255, 255, ...`,
- Unconditional wrapping: `0, 1, ..., 254, 255, 0, 1, ...`,
- Unconditional panic "off-by-one": `0, 1, ..., 254, panic, ...`
- Debug-conditional panic "off-by-one" or wrapping, applied identically to `nth` as to `next`, or
- Through some amazing hack, add an extra bit of state to `RangeFrom` without breaking public API.

The last is impossible without drastically changing how struct initialization works in Rust, so is a no-go.

The unconditional "off-by-one" panic is consistent and simple, and is consistent with interpreting `n..` as `n..MAX` (not `n..=MAX`). Still, this feels like an incorrect choice semantically to me. This also makes e.g. `(0u8..).take(256)` not work, even though iteration is capped before overflow, requiring instead use of `0u8..=255u8` (which to be fair is actually less state, `2*u8` rather than `u8+usize`).

Unconditional saturation, unconditional wrapping, and debug-conditional are "just" a choice of semantics, though the former two would break the debug-checked custom of Rust's integer types. However, it's actually a bit more complicated than that, because of `nth`. `successor_wrapping`, `forward_wrapping`, and `forward_saturating` aren't (efficiently) implementable by this minimum required API surface (of `forward_checked(n)` and `backward_checked(n)`). If we want either of these semantics, we'd have to add a batch operation for it to `Step` in order to keep `RangeFrom::nth` O(1).

Writing all of this out, I think I'm now leaning towards renaming `forward` to `forward_checked`, and adding a `forward` that has the same "panic or wrap" semantics as `successor`.

Reply:
Or, equivalently, giving `successor` (and friends) a step parameter as well.